### PR TITLE
Jbide 21857: Hot class reload doesn't work on OpenShift

### PR DESCRIPTION
**Pull request by tsmaeder**
Currently, this fix only works with the EAP 6.4 Openshift image. If approved, we can add more client libraries

---

This change is [Reviewable](https://reviewable.io/reviews/jbosstools/jbosstools-openshift/1175)

### CODE REVIEW

File: plugins/org.jboss.tools.jmx.remoting1_1_3/.classpath [hunk 1]
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (adietish) [inline plugins/org.jboss.tools.jmx.remoting1_1_3/.classpath]: rather weird plugin name? not sure if we really want to include the version in the plugin name since we'd have to update the plugin name if we have to upgrade the jmx library rthat's included.

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/ConnectorProviderRegistry.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
+package org.jboss.tools.openshift.core.jmx;
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/ConnectorProviderRegistry.java]: include class header please

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/OpenShiftServerAdapter.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
+package org.jboss.tools.openshift.core.jmx;
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/OpenShiftServerAdapter.java]: include class header please

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/OpenshiftConnectionProvider.java [hunk 4]
```diff
@@ -0,0 +1,147 @@
+package org.jboss.tools.openshift.core.jmx;
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/OpenshiftConnectionProvider.java]: include class header please

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/VersionDetector.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
+package org.jboss.tools.openshift.core.jmx;
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/VersionDetector.java]: include class header please

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/ServerType.java [hunk 6]
```diff
@@ -0,0 +1,25 @@
+package org.jboss.tools.openshift.core.jmx;
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/ServerType.java]: include class header please

File: plugins/org.jboss.tools.jmx.remoting1_1_3/build.properties [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+               lib/jboss-logmanager-1.5.3.Final.jar,\
+               lib/jboss-marshalling-1.4.8.Final.jar,\
+               lib/jboss-marshalling-river-1.4.8.Final.jar,\
+               lib/jboss-remoting-3.3.3.Final.jar,\
+               lib/jboss-sasl-1.0.4.Final.jar,\
+               lib/jdeparser-1.0.0.Final.jar,\
+               lib/remoting-jmx-1.1.3.Final.jar,\
+               lib/xnio-api-3.0.10.GA.jar,\
+               lib/xnio-nio-3.0.10.GA.jar,\
+               plugin.xml
```
Comment (adietish) [inline plugins/org.jboss.tools.jmx.remoting1_1_3/build.properties]: dont think that we should commit all these binaries since they should (and jmx already is) fetched and copied via maven dependency plugin

File: plugins/org.jboss.tools.jmx.remoting1_1_3/pom.xml [hunk 8]
```diff
@@ -0,0 +1,59 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+	<modelVersion>4.0.0</modelVersion>
+	<groupId>org.jboss.tools.jmx.remoting1_1_3</groupId>
+	<artifactId>org.jboss.tools.jmx.remoting1_1_3</artifactId>
```
Comment (adietish) [inline plugins/org.jboss.tools.jmx.remoting1_1_3/pom.xml]: want to include all binaries that we use so that they all get fetched and copied via maven dependency plugin?

File: features/org.jboss.tools.openshift.express.feature/.project [hunk 9]
```diff
@@ -10,8 +10,14 @@
...
 			</arguments>
 		</buildCommand>
+		<buildCommand>
+			<name>org.eclipse.m2e.core.maven2Builder</name>
+			<arguments>
+			</arguments>
+		</buildCommand>
 	</buildSpec>
 	<natures>
+		<nature>org.eclipse.m2e.core.maven2Nature</nature>
```
Comment (adietish) [inline features/org.jboss.tools.openshift.express.feature/.project]: this looks unrelated

File: features/org.jboss.tools.openshift.express.test.feature/.project [hunk 10]
```diff
@@ -10,8 +10,14 @@
...
 			</arguments>
 		</buildCommand>
+		<buildCommand>
+			<name>org.eclipse.m2e.core.maven2Builder</name>
+			<arguments>
+			</arguments>
+		</buildCommand>
 	</buildSpec>
 	<natures>
+		<nature>org.eclipse.m2e.core.maven2Nature</nature>
```
Comment (adietish) [inline features/org.jboss.tools.openshift.express.test.feature/.project]: this looks unrelated, should be in a separate PR I think

File: tests/org.jboss.tools.openshift.express.test/.project [hunk 11]
```diff
@@ -20,8 +20,14 @@
...
 			</arguments>
 		</buildCommand>
+		<buildCommand>
+			<name>org.eclipse.m2e.core.maven2Builder</name>
+			<arguments>
+			</arguments>
+		</buildCommand>
 	</buildSpec>
 	<natures>
+		<nature>org.eclipse.m2e.core.maven2Nature</nature>
```
Comment (adietish) [inline tests/org.jboss.tools.openshift.express.test/.project]: this looks unrelated, should be in a separate PR I think

File: plugins/org.jboss.tools.openshift.core/.classpath [hunk 12]
```diff
@@ -1,7 +1,7 @@
...
-	<classpathentry kind="con" path="org.eclipse.pde.core.requiredPlugins"/>
-	<classpathentry kind="src" path="src"/>
-	<classpathentry kind="output" path="bin"/>
-</classpath>
+<?xml version="1.0" encoding="UTF-8"?>
+<classpath>
+	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.8"/>
+	<classpathentry kind="con" path="org.eclipse.pde.core.requiredPlugins"/>
+	<classpathentry kind="src" path="src"/>
+	<classpathentry kind="output" path="target/classes"/>
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.core/.classpath]: this looks unrelated, should be in a separate PR I think

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/OpenshiftConnectionProvider.java [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+		}
+		return null;
+	}
+
+	private IConnectionWrapper createConnection(IServer server) throws MalformedURLException {
+		ConnectorProviderRegistry connectorProviders = OpenShiftCoreActivator.getDefault().getConnectorProviders();
+		VersionKey version = new VersionDetector(server).guess();
+		if (version != null) {
+			JMXConnectorProvider provider = connectorProviders.getProvider(version);
+			return new OpenshiftConnectionWrapper(this, server, provider);
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/OpenshiftConnectionProvider.java]: what if provider == null?

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/OpenshiftConnectionProvider.java [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+	}
+
+	@Override
+	public boolean canDelete(IConnectionWrapper wrapper) {
+		return false;
+	}
+
+	@Override
+	public IConnectionWrapper createConnection(@SuppressWarnings("rawtypes") Map map) throws CoreException {
+		throw new UnsupportedOperationException("Cannot create connection");
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/OpenshiftConnectionProvider.java]: cant implement this (what about the #createConnection that we already have, which is close)? no added value?

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/OpenshiftConnectionWrapper.java [hunk 15]
```diff
@@ -0,0 +1,150 @@
+/*******************************************************************************
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/OpenshiftConnectionWrapper.java]: wrong header?

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/OpenshiftConnectionWrapper.java [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+	}
+
+	public boolean canControl() {
+		return true;
+	}
+
+	public synchronized void connect() throws IOException {
+		try {
+			IDeploymentConfig dc = OpenShiftServerUtils.getDeploymentConfig(server);
+			DebuggingContext debugContext = OpenShiftDebugUtils.get().getDebuggingContext(dc);
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/OpenshiftConnectionWrapper.java]: what if dc == null?

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/OpenshiftConnectionWrapper.java [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+			HashMap<String, Object> env = new HashMap<>();
+			env.put(JMXConnector.CREDENTIALS,
+					new String[] { debugContext.getAdminUsername(), debugContext.getAdminPassword() });
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/OpenshiftConnectionWrapper.java]: what if debugContext == null?

Comment (adietish) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/OpenshiftConnectionWrapper.java]: to me line 76 - 81 look like an ideal candidate for a method #getFirstPod(server)

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/OpenshiftConnectionWrapper.java [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+			IPod firstPod = OpenShiftDebugUtils.get().getFirstPod(dc);
+			if (firstPod != null) {
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/OpenshiftConnectionWrapper.java]: to me line 76 - 81 look like an ideal candidate for a method #getFirstPod(server)

Comment (adietish) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/OpenshiftConnectionWrapper.java]: what if no "firstPod" was found? Shouldnt we tell the user?

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/OpenshiftConnectionWrapper.java [hunk 19]
```diff
@@ -0,0 +1,150 @@
...
+			}
+		} catch (CoreException e) {
+			throw new IOException(e);
+		}
+	}
+
+	public synchronized void disconnect() throws IOException {
+		// close
+		root = null;
+		isConnected = false;
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/OpenshiftConnectionWrapper.java]: shouldnt we set isConnected to false only once we successfully closed the connection?

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/VersionDetector.java [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+	private static final Map<String, VersionKey> knowImages= buildVersionMap();
+	
+	private IServer server;
+	public VersionDetector(IServer server) {
+		this.server= server;
+	}
+	
+	private static Map<String, VersionKey> buildVersionMap() {
+		Map<String, VersionKey> result= new HashMap<String, VersionKey>();
+		result.put("jboss-eap64-openshift", new VersionKey(ServerType.EAP, "6.4.0"));
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/VersionDetector.java]: could we please extract these strings to a constant for easier maintenance?

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/VersionDetector.java [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+		result.put("jboss-eap64-openshift", new VersionKey(ServerType.EAP, "6.4.0"));	
+		return result;
+	}
+
+	VersionKey guess() {
+		IService service = OpenShiftServerUtils.getService(server);
+		if (service != null) {
+			IProject project = service.getProject();
+			if (project != null) {
+				List<IBuildConfig> buildConfigs = project.getResources(ResourceKind.BUILD_CONFIG);
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/VersionDetector.java]: long running operation. We should at least mention this in the javadoc.

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/VersionDetector.java [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+				if (buildConfigs != null) {
+					IBuildConfig buildConfig = ResourceUtils.getBuildConfigForService(service, buildConfigs);
+					DockerImageURI imageUri = OpenShiftServerUtils.getImageUri(buildConfig);
+					return knowImages.get(imageUri.getName());
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/VersionDetector.java]: what if imageUri == null?

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/VersionKey.java [hunk 23]
```diff
@@ -0,0 +1,33 @@
+package org.jboss.tools.openshift.core.jmx;
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/VersionKey.java]: please add class headers

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/server/OpenShiftServerUtils.java [hunk 24]
```diff
@@ -501,6 +492,21 @@ public static boolean isEapStyle(IBuildConfig buildConfig) {
 		}
 		return isEapStyle;
 	}
+
+	public static DockerImageURI getImageUri(IBuildConfig buildConfig) {
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/server/OpenShiftServerUtils.java]: move this to ResourceUtils?

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/server/behavior/OpenShiftLaunchController.java [hunk 25]
```diff
@@ -284,7 +284,9 @@ private int mapPortForwarding(final DebuggingContext debuggingContext, final IPr
 			if(monitor.isCanceled()) {
 				return -1;
 			}
-			port.setLocalPort(SocketUtil.findFreePort());
+			if (port.getRemotePort() != 9999) {
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/server/behavior/OpenShiftLaunchController.java]: can we pelase extract this to a constant or reuse one that already exists?

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/OpenshiftConnectionWrapper.java [hunk 26]
```diff
@@ -0,0 +1,150 @@
...
+			throw new JMXException(s);
+		}
+	}
+
+	public void run(IJMXRunnable runnable, HashMap<String, String> prefs) throws JMXException {
+		run(runnable);
+	}
+
+	private int getLocalAdminPort(IPod pod) {
+		return PortForwardingUtils.getForwardablePorts(pod).stream().filter(p -> p.getRemotePort() == 9999)
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/OpenshiftConnectionWrapper.java]: can we pelase extract this to a constant (in OpenShiftDebugUtils?) and (re)use in
* here
* OpenShiftLaunchController
* OpenShiftDebugUtils

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/server/debug/OpenShiftDebugUtils.java [hunk 27]
```diff
@@ -145,44 +151,76 @@ public void done(IJobChangeEvent event) {
...
+				debugContext.setAdminPassword("adminpw$"+new Random().nextInt(Integer.MAX_VALUE));
+			}
+			deploymentConfig.setEnvironmentVariable(ADMIN_USERNAME_KEY, debugContext.getAdminUsername());
+			deploymentConfig.setEnvironmentVariable(ADMIN_PASSSWD_KEY, debugContext.getAdminPassword());
+		}
+	}
+
+	private boolean updateJMXPort(DebuggingContext debugContext, Set<IPort> ports) {
+		IPort existing = ports.stream().filter(p->p.getName().equals("remotingjmx")).findFirst().orElse(null);
+		PortSpecAdapter newPort = new PortSpecAdapter("remotingjmx", "TCP", 9999);
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/server/debug/OpenShiftDebugUtils.java]: can we pelase define and reuse a constant for this port? (see above)

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/server/debug/OpenShiftDebugUtils.java [hunk 28]
```diff
@@ -145,44 +151,76 @@ public void done(IJobChangeEvent event) {
...
+			if (debugContext.getAdminPassword() == null) {
+				debugContext.setAdminPassword("adminpw$"+new Random().nextInt(Integer.MAX_VALUE));
+			}
+			deploymentConfig.setEnvironmentVariable(ADMIN_USERNAME_KEY, debugContext.getAdminUsername());
+			deploymentConfig.setEnvironmentVariable(ADMIN_PASSSWD_KEY, debugContext.getAdminPassword());
+		}
+	}
+
+	private boolean updateJMXPort(DebuggingContext debugContext, Set<IPort> ports) {
+		IPort existing = ports.stream().filter(p->p.getName().equals("remotingjmx")).findFirst().orElse(null);
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/server/debug/OpenShiftDebugUtils.java]: can we please extract "remotingjmx" to a constant for easier maintenance?

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/server/debug/OpenShiftDebugUtils.java [hunk 29]
```diff
@@ -145,44 +151,76 @@ public void done(IJobChangeEvent event) {
...
+			ports.add(newPort);
+			return true;
+		}
+	}
+
+	private boolean updateDebugPort(DebuggingContext debugContext, Set<IPort> ports) {
+		IPort existing = ports.stream().filter(p -> p.getContainerPort() == debugContext.getDebugPort()).findFirst()
+				.orElse(null);
+		if (existing == null) {
+			return ports.add(new PortSpecAdapter("debug", "TCP", debugContext.getDebugPort()));
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/server/debug/OpenShiftDebugUtils.java]: constant?

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/server/debug/OpenShiftDebugUtils.java [hunk 30]
```diff
@@ -205,92 +243,96 @@ public DebuggingContext getDebuggingContext(IDeploymentConfig deploymentConfig)
...
 			return envVar.get().getValue();
 		}
 		return null;
 	}
-	
+
 	public IPod getFirstPod(IDeploymentConfig dc) {
-		IPod pod = ResourceUtils.getPodsForDeploymentConfig(dc).stream().findFirst() 
-				.orElse(null);
+		IPod pod = ResourceUtils.getPodsForDeploymentConfig(dc).stream().findFirst().orElse(null);
```
Comment (adietish) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/server/debug/OpenShiftDebugUtils.java]: what if ResourceUtils.getPodsForDeploymentConfig(dc) return null?

File: plugins/org.jboss.tools.jmx.remoting1_1_3/.classpath (hunk 1 above)
Comment (tsmaeder) [inline plugins/org.jboss.tools.jmx.remoting1_1_3/.classpath]: The idea here is to have a bundle per version of the client library we want to use. If we want to use 2.0, we will add a bundle with those libraries, not update the this one.

File: plugins/org.jboss.tools.jmx.remoting1_1_3/pom.xml (hunk 8 above)
Comment (tsmaeder) [inline plugins/org.jboss.tools.jmx.remoting1_1_3/pom.xml]: If you can tell me how to do that?

File: plugins/org.jboss.tools.jmx.remoting1_1_3/.classpath (hunk 1 above)
Comment (maxandersen) [inline plugins/org.jboss.tools.jmx.remoting1_1_3/.classpath]: isnt' these not already in server as jars anyway ? seems we should make sure this gets done in a way so they can be reused instead of we keep adding jboss AS jars runtime jars.

Comment (tsmaeder) [inline plugins/org.jboss.tools.jmx.remoting1_1_3/.classpath]: No, the "main" way to load the client jars is to dynamically load them from an installation on disk. According to @robstryker this is done because not all client jars work with all versions of wildfly/eap. If this is only about supporting EAP images on openshift, I guess a single version (latest stable) should do and could be testable.

File: features/org.jboss.tools.openshift.express.feature/.project (hunk 9 above)
Comment (tsmaeder) [inline features/org.jboss.tools.openshift.express.feature/.project]: Yes, but my workspace does not compile without it. Neither does yours, I would think.

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/OpenshiftConnectionProvider.java (hunk 14 above)
Comment (tsmaeder) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/OpenshiftConnectionProvider.java]: This is the same as for the other connection providers for jboss servers. They are managed automatically be the server adapter, basically (see AbstractJBossJMXConnectionProvider)

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/OpenshiftConnectionWrapper.java (hunk 15 above)
Comment (tsmaeder) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/OpenshiftConnectionWrapper.java]: Kind of. I copied the impelementation. I will adapt the header to reflect my contribution.

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/OpenshiftConnectionWrapper.java (hunk 16 above)
Comment (tsmaeder) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/OpenshiftConnectionWrapper.java]: We crash and burn.

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/OpenshiftConnectionWrapper.java (hunk 17 above)
Comment (tsmaeder) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/OpenshiftConnectionWrapper.java]: will only happen if dc == null.

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/OpenshiftConnectionWrapper.java (hunk 18 above)
Comment (tsmaeder) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/OpenshiftConnectionWrapper.java]: I don't think so. We are not in userland here. If anything, I guess we should throw an IOException to signal connection failure. However, other implementation also fail without throwing (see JBossServerConnection, for example)

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/OpenshiftConnectionWrapper.java (hunk 19 above)
Comment (tsmaeder) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/OpenshiftConnectionWrapper.java]: Not sure...the javadoce for the connector says:

@exception IOException if the connection cannot be closed
     * cleanly.  If this exception is thrown, it is not known whether
     * the server end of the connection has been cleanly closed.

I think assuming that the connector is closed is the safer assumption.

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/VersionDetector.java (hunk 21 above)
Comment (tsmaeder) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/VersionDetector.java]: Yes, creating the connection as a whole should happen as a background job.

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/VersionDetector.java (hunk 22 above)
Comment (tsmaeder) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/VersionDetector.java]: hashmap handles null key.

Comment (tsmaeder) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/VersionDetector.java]: but the build config may be null

Comment (tsmaeder) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/VersionDetector.java]: crash and burn...

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/server/OpenShiftServerUtils.java (hunk 24 above)
Comment (tsmaeder) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/server/OpenShiftServerUtils.java]: Yes.

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/VersionDetector.java (hunk 20 above)
Comment (tsmaeder) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/jmx/VersionDetector.java]: These strings don't mean anything beyond their literal value and they should not be referenced anywhere outside this method so I don't think a constant is appropriate. Unless we want to have a registry of well known image names, but then it should not live hidden inside this class.

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/server/debug/OpenShiftDebugUtils.java (hunk 30 above)
Comment (tsmaeder) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/server/debug/OpenShiftDebugUtils.java]: it doesn't

**Comment by robstryker** [comment]
This PR is not a proper solution but I did steal some of the work for my other PRs. 

https://github.com/jbosstools/jbosstools-server/pull/451
https://github.com/jbosstools/jbosstools-openshift/pull/1363

The openshift PR depends on the server PR
